packages/libp2p/src/connection-monitor.ts

```diff
@@ -0,0 +1,127 @@
+import type { AdaptiveTimeoutInit } from '@libp2p/utils/adaptive-timeout'
+
+const DEFAULT_PING_INTERVAL_MS = 10000
+
+export interface ConnectionMonitorInit {
+  /**
+   * How often to ping remote peers in ms
+   *
+   * @default 10000
+   */
+  pingInterval?: number
+
+  /**
+   * Timeout settings for How long the ping is allowed to take before the
+   * connection will be judged inactive and aborted.
+   *
+   * The timeout is adaptive to cope with slower networks or nodes that
+   * have changing network characteristics, such as mobile.
+   */
+  pingTimeout?: Omit<AdaptiveTimeoutInit, 'metricsName' | 'metrics'>
+
+  /**
+   * If true, any connection that fails the ping will be aborted
+   *
+   * @default true
+   */
+  abortConnectionOnPingFailure?: boolean
+}
+
+export interface ConnectionMonitorComponents {
+  logger: ComponentLogger
+  connectionManager: ConnectionManager
+  metrics?: Metrics
+}
+
+export class ConnectionMonitor implements Startable {
+  private readonly components: ConnectionMonitorComponents
+  private readonly log: Logger
+  private heartbeatInterval?: ReturnType<typeof setInterval>
+  private readonly pingIntervalMs: number
+  private abortController?: AbortController
+  private readonly timeout: AdaptiveTimeout
+
+  constructor (components: ConnectionMonitorComponents, init: ConnectionMonitorInit = {}) {
+    this.components = components
+
+    this.log = components.logger.forComponent('libp2p:connection-monitor')
+    this.pingIntervalMs = init.pingInterval ?? DEFAULT_PING_INTERVAL_MS
+
+    this.timeout = new AdaptiveTimeout({
+      ...(init.pingTimeout ?? {}),
+      metrics: components.metrics,
+      metricName: 'libp2p_connection_monitor_ping_time_milliseconds'
+    })
+  }
+
+  readonly [Symbol.toStringTag] = '@libp2p/connection-monitor'
+
+  readonly [serviceCapabilities]: string[] = [
+    '@libp2p/connection-monitor'
+  ]
+
+  start (): void {
+    this.abortController = new AbortController()
+
+    this.heartbeatInterval = setInterval(() => {
+      this.components.connectionManager.getConnections().forEach(conn => {
+        Promise.resolve().then(async () => {
+          let start = Date.now()
+          try {
+            const signal = this.timeout.getTimeoutSignal({
+              signal: this.abortController?.signal
+            })
+            const stream = await conn.newStream('/ipfs/ping/1.0.0', {
+              signal,
+              runOnTransientConnection: true
+            })
+            const bs = byteStream(stream)
+            start = Date.now()
+
+            await Promise.all([
+              bs.write(new Uint8Array(1), {
+                signal
+              }),
+              bs.read(1, {
+                signal
+              })
+            ])
+
+            conn.rtt = Date.now() - start
+
+            await bs.unwrap().close({
+              signal
+            })
+          } catch (err: any) {
+            if (err.code !== 'ERR_UNSUPPORTED_PROTOCOL') {
```

Comment:
If we are will to monitor connections that don't support the protocol, should we also include the `ERR_CONNECTION_DENIED` error as well? In a scenario should the connection fail to upgrade?

Reply:
`ERR_CONNECTION_DENIED` is thrown when the remote has exceeded a rate limit and results in the connection being closed so I don't think it's relevant here.